### PR TITLE
NodeMaterial: WebGPU

### PR DESCRIPTION
**Pull request by sunag**
@Mugen87 @mrdoob Very, very draft to facilitate the understanding... this is a step implement to WebGL too, is my new approach, merger node slots directly in ShaderLib...

I fork the example `webgpu_sandbox.html` and added just some lines code:
```js
import FloatNode from './jsm/renderers/nodes/inputs/FloatNode.js';
import Vector3Node from './jsm/renderers/nodes/inputs/Vector3Node.js';
import OperatorNode from './jsm/renderers/nodes/math/OperatorNode.js';

materialBox.colorNode = new OperatorNode( '+', new FloatNode( .5 ).setConst( true ), new Vector3Node( new THREE.Vector3( 0, 0, 1 ) ) );
materialBox.opacityNode = new FloatNode( .5 );
//materialBox.transparent = true;
```

### CODE REVIEW

File: examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
+import { NodeBuilder } from '../../nodes/core/NodeBuilder.js';
```
Comment (Mugen87) [inline examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js:1]: I guess it would be good to avoid all dependencies to `examples/jsm/nodes/`. The idea was start with a fresh system (and only with the minimal bits) which are required for a basic material.

It seems there are dependencies to WebGL related logic in `NodeBuilder` which should not be part of the new system. I also suppose that certain parts of the node material system will look different that what is existing now in `examples/jsm/nodes/` when devs contribute to it. Hence, it seems better starting from scratch in `examples/jsm/renderers/webgpu/node/` with the minimal setup.

Comment (sunag) [reply examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js:1]: > The idea was start with a fresh system (and only with the minimal bits) which are required for a basic material.

Yes, in fact, just this still not ready.. only however is that we have an opening that I would like take advantage... I think in creating an `examples/jsm/renderers/nodes` new `NodeMaterial` system for hybrid renderers and a NodeBuilder for each Renderer, for example... `..renderers/webgl/WebGLNodeBuilder` and `..renderers/webgpu/WebGPUNodeBuilder` extended of `renderers/nodes/core/NodeBuilder`. Any single `Node` can be used in different renderers and WebGL versions. Part of the work I will do the disassociation of WebGL and cleanup.

Comment (sunag) [reply examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js:1]: I will try to be as minimal as possible...

Comment (Mugen87) [reply examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js:1]: It's a meaningful goal to have a rendering agnostic description of materials. So the a node composition itself should have no WebGL or WebGPU associated dependencies. Instead, specific builders produce the respective WebGL or WebGPU output.

Comment (sunag) [reply examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js:1]: I think this the idea but that is no ready I ask you to ignore this version of NodeBuilder for the time being.

Comment (sunag) [reply examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js:1]: A new version of NodeBuilder going to be moved to `examples/jsm/renderers/nodes` for have no WebGL or WebGPU dependencies as the comments above... until the release of this PR.

>  I think in creating an examples/jsm/renderers/nodes new NodeMaterial system for hybrid renderers and a NodeBuilder for each Renderer, for example... ..renderers/webgl/WebGLNodeBuilder and ..renderers/webgpu/WebGPUNodeBuilder extended of renderers/nodes/core/NodeBuilder

**Comment by sunag** [comment]
@Mugen87 folow ( https://github.com/mrdoob/three.js/pull/20421#discussion_r495488625 ) ( https://github.com/mrdoob/three.js/pull/20421/commits/f738e6a89f809a94e78b1ea178e81cc7e1bbfafe )
`/jsm/renderers/nodes` agnostic renderer node system
`/jsm/renderers/webgpu/WebGPUNodeBuilder` WebGPU node builder

File: examples/webgpu_sandbox_nodes.html [hunk 2]
```diff
@@ -51,7 +52,7 @@
 				const geometryBox = new THREE.BoxBufferGeometry();
 				const materialBox = new THREE.MeshBasicMaterial( { map: texture } );
 
-				materialBox.color = new Nodes.ColorNode( 0x0099FF ).setReadonly( true );
+				materialBox.color = new OperatorNode( '+', new Vector3Node( new THREE.Vector3( 0, .6, 0 ) ), new Vector3Node( new THREE.Vector3( 0, 0, 1 ) ) );
```
Comment (sunag) [inline examples/webgpu_sandbox_nodes.html:55]: I add a single OperatorNode to test the flow...

**Comment by Mugen87** [comment]
I don't think it is necessary to make a new example for nodes anymore. Do you mind modifying the existing `webgpu_sandbox_nodes` example?

File: examples/jsm/renderers/nodes/core/NodeBuilder.js [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+		
+	}
+	
+	format( code, fromType, toType ) {
+
+		const typeToType = toType + ' <- ' + fromType;
+
+		switch ( typeToType ) {
+
+			case 'f <- v2' : return code + '.x';
```
Comment (Mugen87) [inline examples/jsm/renderers/nodes/core/NodeBuilder.js:122]: This code section is probably easier to read if you following this pattern: `from -> to`. So:
```js
const conversion = fromType + ' -> ' + toType;

switch ( conversion ) {

    case 'v2 -> f' : return code + '.x';
```

File: examples/jsm/renderers/nodes/core/InputNode.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import Node from './Node.js';
+
+class InputNode extends Node {
+
+	constructor( type ) {
+		
+		super( type );
+		
+		this.isInputNode = true;
```
Comment (Mugen87) [inline examples/jsm/renderers/nodes/core/InputNode.js:9]: During the class migration, it was decided to write the `is*` properties like so:
```js
Object.defineProperty( this, 'isInputNode', { value: true } );
```

File: examples/jsm/renderers/nodes/core/InputNode.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+		
+		// force constant for now
+		this.constant = true;
```
Comment (Mugen87) [inline examples/jsm/renderers/nodes/core/InputNode.js:12]: I guess when `constant` is set to `false`, it's a dynamic input and thus a uniform, right?

File: examples/jsm/renderers/nodes/core/InputNode.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+	
+	generateConst( builder, output ) {
+		
+		console.warn("Abstract function");
+		
+	}
+	
+	generate( builder, output ) {
+		
+		if ( this.constant ) {
```
Comment (Mugen87) [inline examples/jsm/renderers/nodes/core/InputNode.js:24]: In such cases, we lately tend to make such checks more explicit. Meaning:
```js
if ( this.constant === true ) {
```

File: examples/jsm/renderers/nodes/core/NodeBuilder.js [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+		
+		this.nodesData = new WeakMap();
+		
+		this.shader = undefined;
+		
+	}
+	
+	setMaterial( material ) {
+		
+		this.material = material;
```
Comment (Mugen87) [inline examples/jsm/renderers/nodes/core/NodeBuilder.js:18]: Please define `material` in the constructor, too. That makes it easier to understand what properties the class is working with.

File: examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+	}
+	
+	buildShader( shader, code ) {
+		
+		// use regex maybe for security?
+		const versionStrIndex = code.indexOf("\n");
+		
+		let shaderCode = code.substr( 0, versionStrIndex ) + "\n";
+
+		let shaderData = this.build( shader );
```
Comment (Mugen87) [inline examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js:19]: `shaderData` can be defined as `const`.

**Comment by sunag** [comment]
@Mugen87 You have any idea for that error message I am getting:
If I add a single `vec3` in `nodeUniforms` it works done. 24 bytes seems the correct size for two vec3.

image

File: examples/jsm/renderers/nodes/core/NodeBuilder.js [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+		
+		
+	}
+	/*
+	analyzeNode( node ) {
+		
+		
+	}
+	*/
+	flowNode( node, output ) {
```
Comment (Mugen87) [inline examples/jsm/renderers/nodes/core/NodeBuilder.js:63]: Since `output` can be `undefined`, it's easier to read if `output` has a default value.
```js
flowNode( node, output === null ) {
```
We lately use `null` instead of `undefined` since it's more robust to test for `null` instead of relying on `undefined` values.

File: examples/jsm/renderers/nodes/inputs/Vector3Node.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+	constructor( value ) {
+
+		super( 'v3' );
+
+		this.value = value
+
+	}
+
+	generateConst( builder ) {
```
Comment (Mugen87) [inline examples/jsm/renderers/nodes/inputs/Vector3Node.js:13]: For clarity, `output` should added here, too.
```js
generateConst( builder, output = null ) {
```

File: examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+		
+		shaderCode += shaderData.defines;
+
+		shaderCode += code.substr( versionStrIndex );
+
+		console.log( shaderCode );
```
Comment (Mugen87) [inline examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js:25]: For debugging, it's okay to keep this for now.

**Comment by Mugen87** [comment]
> You have any idea for that error message I am getting:

This message did not pop up in my tests so far but I assume that `vec3`s must start on a 16-byte boundary. Can you please change the `byteLength` of `Vector3Uniform` to `16`?

**Comment by Mugen87** [comment]
@mrdoob This PR looks really great! In context of design, I would highlight one important aspect since it will accompany  node material from the very beginning.

With the current implementation, nodes have the responsibility to control their build/generation. This process is mainly done with a node builder with is injected as a parameter. One could consider a different architecture such that nodes do not have any build/generate methods but just describe themselves. You then have a centralized entity that traverses through the node composition and performs the build.

I personally favor the current approach since it's similar to having `toJSON()` and `fromJSON()` on object level instead of a single complex component that handles the entire serialization/deserialization. It's easier to encapsulate code and hide component-specific logic (in this case node specific).

I just want to make sure everybody is aware of this detail.

File: examples/jsm/renderers/nodes/core/NodeBuilder.js [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+class NodeBuilder {
+
+	constructor() {
+		
+		this.slots = { vertex: [], fragment: [] };
+		this.defines = { vertex: {}, fragment: {} };
+		
+		this.nodesData = new WeakMap();
+		
+		this.shader = undefined;
```
Comment (Mugen87) [inline examples/jsm/renderers/nodes/core/NodeBuilder.js:12]: Please use `null` here.

**Comment by Kangz** [comment]
> @Mugen87 You have any idea for that error message I am getting:
> If I add a single `vec3` in `nodeUniforms` it works done. 24 bytes seems the correct size for two vec3.
> 
> image

Mmmmh it should work with the std140 layout, but WebGPU seems to think it is 28 bytes probably because of a bad array stride? Could you share permalink of the GLSL->SPIRV compilation in http://shader-playground.timjones.io/ ?

**Comment by Mugen87** [comment]
I don't know how to share a permalink. Here is some sample code:
```
#version 450
layout(set = 0, binding = 2) uniform MaterialUniforms {
    float opacity;
    vec3 color1;
    vec3 color2;
} materialUniforms;

layout(set = 0, binding = 3) uniform sampler mySampler;
layout(set = 0, binding = 4) uniform texture2D myTexture;

layout(location = 0) in vec2 vUv;
layout(location = 0) out vec4 outColor;

void main() {
    outColor = texture( sampler2D( myTexture, mySampler ), vUv );
    outColor.a *= materialUniforms.opacity;
}
```
It seems that even if a single float only takes 4 bytes, it's necessary to put 12 bytes as padding when starting with `color1`.

**Comment by Mugen87** [comment]
I guess we need the logic from the UBO PR that determines the total buffer size according to the STD140 layout.

https://github.com/mrdoob/three.js/pull/15562/files#diff-cba50499b84f8d3f3690b625a907ba74R209

**Comment by Kangz** [comment]
Ah yes, I misread. It is expect that the buffer would use 28 bytes in std140 layout because things are mostly padded to 16 bytes all the time. Like you said, the logic in the UBO PR should help.

**Comment by Mugen87** [comment]
Okay, added: 1cf4431f75fd163804bc67fee7460d4fcdeb61ef

@sunag After rebasing the mentioned warnings should be gone.

File: examples/jsm/renderers/webgpu/WebGPURenderPipelines.js [hunk 13]
```diff
@@ -808,11 +818,19 @@ const ShaderLib = {
 		layout(set = 0, binding = 3) uniform sampler mySampler;
 		layout(set = 0, binding = 4) uniform texture2D myTexture;
 
+		#pragma node_uniforms
```
Comment (takahirox) [inline examples/jsm/renderers/webgpu/WebGPURenderPipelines.js:821]: Sorry if I miss something. What's this pragma for?

**Comment by sunag** [comment]
> @sunag After rebasing the mentioned warnings should be gone.

@Mugen87 It work now. Thanks guys for quick fix.

File: examples/jsm/renderers/webgpu/WebGPURenderPipelines.js (hunk 13 above)
Comment (sunag) [reply examples/jsm/renderers/webgpu/WebGPURenderPipelines.js:821]: For nothing in this moment... I will replace this for NODE_UNIFORMS like this code: https://github.com/mrdoob/three.js/pull/20421#issuecomment-699906186

File: examples/jsm/renderers/nodes/core/InputNode.js (hunk 5 above)
Comment (sunag) [reply examples/jsm/renderers/nodes/core/InputNode.js:12]: Yes, after implement uniforms I will set `this.constant = false` as default.

**Comment by sunag** [comment]
I think in changing the properties to `material.colorNode` and `material.opacityNode` because the implementation will be easier including in `WebGL` context.

**Comment by sunag** [comment]
I add Nodes support in PointsMaterial ( https://github.com/mrdoob/three.js/pull/20421/commits/99cced52d11222fa898f8740bbccef1ac7834651) in some minutes, very simple... This can be used for any material of threejs too.

File: examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+			
+			if (uniformsGroup === undefined) {
+				
+				uniformsGroup = new WebGPUNodeUniformsGroup( shaderStage );
+				
+				this.uniformsGroup[shaderStage] = uniformsGroup;
+				
+			}
+			
+			let uniformGroup;
```
Comment (Mugen87) [inline examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js:105]: This variable should be called just `uniform`.

File: examples/jsm/renderers/webgpu/WebGPURenderPipelines.js [hunk 15]
```diff
@@ -19,6 +21,8 @@ class WebGPURenderPipelines {
 		this.glslang = glslang;
 		this.sampleCount = sampleCount;
 
+		this.nodeBindings = new WeakMap();
```
Comment (Mugen87) [inline examples/jsm/renderers/webgpu/WebGPURenderPipelines.js:24]: This can be renamed to just `bindings`.

File: examples/jsm/renderers/webgpu/WebGPURenderPipelines.js [hunk 16]
```diff
@@ -200,6 +212,12 @@ class WebGPURenderPipelines {
 
 	}
 
+	getNodeBindings( object ) {
```
Comment (Mugen87) [inline examples/jsm/renderers/webgpu/WebGPURenderPipelines.js:215]: And this to `getBindings()`

File: examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js [hunk 17]
```diff
@@ -0,0 +1,164 @@
+import WebGPUUniformsGroup from './WebGPUUniformsGroup.js';
+import { FloatUniform, Vector3Uniform } from './WebGPUUniform.js';
+
+import NodeSlot from '../nodes/core/NodeSlot.js';
+import NodeBuilder from '../nodes/core/NodeBuilder.js';
+
+class WebGPUNodeUniformsGroup extends WebGPUUniformsGroup {
```
Comment (Mugen87) [inline examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js:7]: Can you also use just `WebGPUUniformsGroup`? Eventually, bindings should only be managed over the node material system.

**Comment by Mugen87** [comment]
Nice! I think we should manage now all existing uniforms over the node material system so the following code block can be removed:

https://github.com/mrdoob/three.js/blob/b1390eb11be57c66b1239b822039198621ecb500/examples/jsm/renderers/webgpu/WebGPUBindings.js#L40-L56

Instead of this code, we just have this:
```js
const bindings = this.pipelines.getBindings( object );
```

**Comment by Mugen87** [comment]
When you look at the shader code in `WebGPURenderPipelines`, you can see that the UBOs `CameraUniforms` and `ModelUniforms` are required in all programs so far. It's essentially a replacement for the pre-defined code in `WebGLProgram`. 

When managing all bindings over the node material system, I suggest to finally review and then merge this PR since it is a proper first step.

File: examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+		if ( uniformsFragmentGroup ) bindings.push( uniformsFragmentGroup );
+		
+		return bindings;
+		
+	}
+	
+	createUniformFromNode( node, shaderStage, type ) {
+		
+		const uniformNode = super.createUniformFromNode( node, shaderStage, type ) 
+		const nodeData = this.createDataFromNode( node, shaderStage );
```
Comment (Mugen87) [inline examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js:91]: This call irritated me a bit. You call `createDataFromNode()` here although this method was already called inside of `createUniformFromNode()`. I understand that you only call it to _get_ the (already existing) node data. 

This code will be more clear when it is align to `WebGPUProperties` or `WebGLProperties`. Meaning to transform it to a getter.  The implementation is already similar to:

https://github.com/mrdoob/three.js/blob/b1390eb11be57c66b1239b822039198621ecb500/src/renderers/webgl/WebGLProperties.js#L5-L18
Meaning the method creates a cached object if necessary. But this creation is not something the caller has not know about. I suggest to use `getUniformNodeData()`. The code will look like so then:
```js
const uniformNode = super.getUniformFromNode( node, shaderStage, type ) 
const nodeData = this.getUniformNodeData( node, shaderStage );
```
In my mind `create*()` methods are usually factory methods that always return a new instance when called. This would is currently not true for both uniform methods.

Comment (sunag) [reply examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js:91]: This was get before `getNodeData` I will keep and I have that change the logic of  `createUniformFromNode`... This code block changed a lot from yesterday to today because of the implementation of texture2d uniform...

**Comment by sunag** [comment]
> When managing all bindings over the node material system, I suggest to finally review and then merge this PR since it is a proper first step.

I am already thinking about cleaning and `Ready for review` this PR after implement `TextureNode`.

**Comment by sunag** [comment]
> Instead of this code, we just have this:
const bindings = this.pipelines.getBindings( object );

It is not clear to me where you want to move these functions, it is for `WebGPURenderPipelines` ?

**Comment by Mugen87** [comment]
> It is not clear to me where you want to move these functions, it is for WebGPURenderPipelines ?

I initially thought we could remove them and add the common bindings (`CameraUniforms` and `ModelUniforms`) with the node material system.

**Comment by sunag** [comment]
> I thought we could remove them and add the common bindings (CameraUniforms and ModelUniforms) with the node material system.

This would change a lot, a very simple implementation would mix the logic of the procedural with natives... I don't think it's appropriate for the moment.

**Comment by Mugen87** [comment]
No problem. We can experiment later with this.

**Comment by sunag** [comment]
If so, I see no reason not to leave the properties hybrid right now because code block like this don't make much sense... thinking ahead.

```js
layout(set = 0, binding = 2) uniform OpacityUniforms {
	float opacity;
} opacityUniforms;

layout(set = 0, binding = 3) uniform sampler mySampler;
layout(set = 0, binding = 4) uniform texture2D myTexture;
```

**Comment by Mugen87** [comment]
Yeah, the current shader code was more or less developed for testing. If we can consolidate this in order to avoid unnecessary bindings we should do so.

**Comment by sunag** [comment]
> We just have to be aware that UBOs might be shared across multiple shader programs. So e.g. CameraUniforms should be a unique UBO which is reused.

Yes, this feature already exists in the current node material system we should only bring it to the new node system: ( https://github.com/mrdoob/three.js/blob/dev/examples/jsm/nodes/accessors/CameraNode.js#L100 )

**Comment by Mugen87** [comment]
@sunag Do you think you can finish the PR in the next time by adding support for `TextureNode`? Otherwise I would give it a try since I've made myself a little bit more familiar with the node material system. 

It would be good to have a basic version of the node material in place so we can move on with developing the renderer at other places.

**Comment by sunag** [comment]
@Mugen87 I'm looking for sponsors, so I can focus on this work.

File: examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+		return bindings;
+
+	}
+
+	getUniformFromNode( node, shaderStage, type ) {
+
+		const uniformNode = super.getUniformFromNode( node, shaderStage, type );
+		const nodeData = this.getDataFromNode( node, shaderStage );
+
+		if ( nodeData.uniformGroup === undefined ) {
```
Comment (takahirox) [inline examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js:92]: Is this check correct? If I'm right, `nodeData.uniformGroup` seems always `undefined` because `.uniformGroup` doesn't seem to be assigned anywhere (`$ find ./examples/jsm/renderers/ -print0 | xargs -0 grep uniformGroup` matches only this line).

**Comment by takahirox** [comment]
It's ok to do it in another PR but I think moving `_parseShaderAttributes()` from `WebGPURenderPipelines` to `WebGPUNodeBuilder` would be good because

1. `WebGPURenderPipelines` will be simpler by being more isolated from the shader code
2. Parsing the shader attributes would be easier because I guess `WebGPUNodeBuilder` may manage not only uniforms but also shader attributes near future

File: examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js (hunk 19 above)
Comment (sunag) [reply examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js]: Fixed! Thanks.

File: examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+		return bindings;
+
+	}
+
+	getUniformFromNode( node, shaderStage, type ) {
+
+		const uniformNode = super.getUniformFromNode( node, shaderStage, type );
+		const nodeData = this.getDataFromNode( node, shaderStage );
+
+		if ( nodeData.uniform === undefined ) {
```
Comment (takahirox) [inline examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js:92]: If I'm right `nodeData.uniform` is always non-`undefined` here because it's set in `super.getUniformFromNode()`. I guess what you really wanted to was

```javascript
if ( nodeData.uniformsGrouop === undefined ) {
  ....
  nodeData.uniformsGroup = uniformsGroup;
}
```

Comment (sunag) [reply examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js]: I rename of `nodeData.uniformGroup` to `nodeData.uniform` because it is not a `group` reviewed by @Mugen87. I renamed to `nodeData.gpuUniform` now but we can put a better name,

**Comment by takahirox** [comment]
Thanks for the update and explanation for `nodeData.webgpuUniform`.

The references `nodeData.uniform` and `nodeData.webgpuUniform` don't seem to be used other than just indicating whether uniforms for a node are initialized or not. I guess you plan to use them somewhere near future so you saved them in `nodeData`.

I would like to suggest to save boolean flags like `(webgpu)uniformIsInitialized` instead until we actually use the saved uniforms somewhere because

1. It'll be easier to understand especially for new node based material system devs like me
2. If we end up not using `nodeData.(webgpu)uniform` and forgetting to remove the references, the references are unnecessarily saved and the memory could be leaked

**Comment by takahirox** [comment]
>@sunag Do you think you can finish the PR in the next time by adding support for TextureNode? Otherwise I would give it a try since I've made myself a little bit more familiar with the node material system.
>
>It would be good to have a basic version of the node material in place so we can move on with developing the renderer at other places.

I would help, too, because I studied Node based material system and our implementation.

**Comment by takahirox** [comment]
> With the current implementation, nodes have the responsibility to control their build/generation. This process is mainly done with a node builder with is injected as a parameter. One could consider a different architecture such that nodes do not have any build/generate methods but just describe themselves. You then have a centralized entity that traverses through the node composition and performs the build.
>
> I personally favor the current approach since it's similar to having toJSON() and fromJSON() on object level instead of a single complex component that handles the entire serialization/deserialization. It's easier to encapsulate code and hide component-specific logic (in this case node specific).

Yeah, the current approach would be one of the good designs.

But I think in the current implementation `Node` may have a bit too strong dependency with `Builder`. For example `FloatNode.generateConst()` calls `builder.generateFloat()` to generate the online shader.

Ideally it would be good if `Node` has less dependency with `Builder` because the design would be simpler. In the above example, `.generateFloat()` might be moved to `Node` or somewhere else. (Sorry I haven't looked into deeply yet tho).

This cleaning up can be done in another PR (if needed).

**Comment by sunag** [comment]
> Ideally it would be good if Node has less dependency with Builder because the implementation would be simpler. In the above example, .generateFloat() might be moved to Node or somewhere else. (Sorry I haven't looked into deeply yet tho).

It is necessary to interpret `NodeBuilder` as a code generator, it is equivalent the a `Emitter` or `Code Generator` of a `Compiler`, everything related to the output programming language should be emit by him.
An example: https://github.com/saharan/OGSL/blob/7727e6a282a57e0103232711d1436c2365d248e8/src/com/element/oimo/ogsl/OGSL.as#L1341

We are not trust that `WebGL3` or another advance uses always `vec3`, maybe can use `float3()` or `assembly` instead of `vec3` in `NodeVector3.generateConst()` for example.  `NodeBuilder` as main `Emitter` will be prepared to compile any node for a specific output without having to change any node, equivalent to the model (JIT). Anyway almost everything could be fixed with a polyfills but I dont't think it would be appropriate if we can avoid. If we also want use a `Optimizer` in `NodeBuilder` for a specific output is important treat as the main emitter.

**Comment by sunag** [comment]
Fact that the PR is a very initial implementation not show other uses for this, but an example more clean. Compilers optimizers make this too...

In case of multiples use of `NodeBuilder.generateFloat( 1.0 )`, `NodeBuilder` can create a `const` and return the variable name like `nodeC0` for example instead of `1.0`.

This is an reason that `analyze` is called before of `build` in `flow` process.

**Comment by takahirox** [comment]
Yes, we will may need Code generator to support multiple languages and/or shader optimization. But I preferred not to include it in the initial PR to make it minimal and simple.

Currently we support only GLSL 4.5 (->SPIRV) and we may switch to WGSL when it will be ready, I'm not sure if we will support GLSL 4.5 or other languages. And we haven't taken the shader optimization into account yet. We can discuss Code generator later instead when we actually need.

But this is just my preference and no strong opinion. Probably we don't need to drop the feature already implemented. And showing the direction of the flexibility to languages and optimization would be good in the initial PR. Please listen my opinion as just my two cents.

**Comment by Mugen87** [comment]
As far as I have understood, when WGSL is ready and browser support it there is no need for supporting GLSL 4.5 anymore. 

@Kangz Is that right? Or will it be required to support two shading languages?

**Comment by takahirox** [comment]
Regarding `TextureNode`, I would like to suggest we implement it in another PR.

Because, if I'm right, `sampler` and `texture2D` can't be in uniform structure so we may need to manage binding numbers and others. It might be a bit complex so if we implement it in another PR, the review would be easier.

**Comment by Mugen87** [comment]
I've hacked in experimental texture support a while ago and the necessary changes were manageable. Although the code required some clean up, the amount of new lines was manageable. 

I think we should start with a node material that can replace the current material usage in `webgpu_sandbox`.

**Comment by takahirox** [comment]
> I vote to start with a node material that can replace the current material usage in webgpu_sandbox.

Yes, that would be ideal to me, too. But @sunag seems to have a very heavy workload of WebGPU node based material system. I know we don't need to be in hurry but I suggested that because I wanted to seek a way to divide the workload among us (and to make the review easier). Of course, no strong opinion.

**Comment by sunag** [comment]
I already start the dev of `TextureNode`. I `commit` in the next days...

image

**Comment by sunag** [comment]
I made a texture animated to test, this `commit` is a WIP.
https://raw.githack.com/sunag/three.js/nodematerial-webgpu/examples/webgpu_sandbox.html

File: examples/jsm/renderers/webgpu/WebGPUUniform.js [hunk 21]
```diff
@@ -19,21 +19,35 @@ class WebGPUUniform {
...
+		return this.value;
+		
+	}
 
 }
 
 class FloatUniform extends WebGPUUniform {
 
-	constructor( name, value = 0 ) {
+	constructor( nodeUniform ) {
```
Comment (Mugen87) [inline examples/jsm/renderers/webgpu/WebGPUUniform.js:33]: I think it's better if the classes that extend `WebGPUUniform` do not know about the node material. Hence, I prefer the previous ctor signature.

File: examples/jsm/renderers/webgpu/WebGPUBindings.js [hunk 22]
```diff
@@ -155,39 +155,31 @@ class WebGPUBindings {
 			} else if ( binding.isSampler ) {
 
 				const material = object.material;
-				const texture = material[ binding.name ];
+				const texture = binding.texture;
 
-				if ( texture !== null ) {
```
Comment (Mugen87) [inline examples/jsm/renderers/webgpu/WebGPUBindings.js:156]: If we consider the texture parameter of `WebGPUSampledTexture` as mandatory, then removing this check is okay. Meaning `WebGPUSampledTexture.texture` can never be `null`.

**Comment by sunag** [comment]
@Mugen87 I move `WebGPUNode*`  to `./nodes` folder in order to facilitate.

File: examples/webgpu_sandbox.html [hunk 23]
```diff
@@ -84,9 +113,14 @@
 
 				}
 
+
 				const geometryPoints = new THREE.BufferGeometry().setFromPoints( points );
 				const materialPoints = new THREE.PointsMaterial();
 
+				geometryPoints.addAttribute( 'color', geometryPoints.getAttribute( 'position' ) );
```
Comment (Mugen87) [inline examples/webgpu_sandbox.html:120]: Nit: Should be `setAttribute()`.

File: examples/jsm/renderers/webgpu/nodes/WebGPUNodes.js [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+
+	}
+
+	get( object ) {
+
+		let data = this.uniformsData.get( object );
+
+		if ( data === undefined ) {
+			
+			this.uniformsData.set( object, data );
```
Comment (Mugen87) [inline examples/jsm/renderers/webgpu/nodes/WebGPUNodes.js:16]: I guess you the following line is missing, otherwise `undefined` is set:
```js
data = {};
```
It seems `WebGPUNodes` is not yet used in the renderer. What are your planes for this class?

Comment (sunag) [reply examples/jsm/renderers/webgpu/nodes/WebGPUNodes.js:16]: Compute `bindings` replacing functions like `WebGPUBindings._getMeshBasicBindings()`, get all `nodes` of a material for update in each `frame` and soon generate `hash` per material...

**Comment by sunag** [comment]
I suggest merge this PR while I doing the others, in case it's ok for you too. I update all WebGPU examples.

**Comment by sunag** [comment]
I think that the next step is move `CameraUniforms` and `ModelUniforms` to `nodes`, right?

**Comment by sunag** [comment]
Examples
https://raw.githack.com/sunag/three.js/nodematerial-webgpu/examples/index.html?q=webgpu#webgpu_sandbox

**Comment by Mugen87** [comment]
> I think that the next step is move CameraUniforms and ModelUniforms to nodes, right?

That sounds good. The renderer should not manage/setup UBOs but the material system. I've initially added this code in `WebGPUBindings` so we have something to start.

**Comment by Mugen87** [comment]
> I suggest merge this PR while I doing the others, in case it's ok for you too.

I would say yes. The general structure of the node material system looks good to me. Merging this PR means we have something on `dev` to experiment with. Further tweaks and changes can be discussed in separate PRs.

**Review by Mugen87 — APPROVED** [review]

File: examples/jsm/renderers/nodes/core/NodeBuilder.js [hunk 25]
```diff
@@ -0,0 +1,320 @@
...
+
+		}
+
+		this.shaderStage = null;
+
+		for ( let shaderStage of shaderStages ) {
+
+			this.define( shaderStage, 'NODE_HEADER_UNIFORMS', this.getUniformsHeaderSnippet( shaderStage ) );
+			this.define( shaderStage, 'NODE_HEADER_ATTRIBUTES', this.getAttributesHeaderSnippet( shaderStage ) );
+			this.define( shaderStage, 'NODE_BODY_ATTRIBUTES', this.getAttributesBodySnippet( shaderStage ) );
```
Comment (Mugen87) [inline examples/jsm/renderers/nodes/core/NodeBuilder.js:305]: When looking at the code, does the distinction of header and body mean?

`NODE_HEADER_ATTRIBUTES`-> attribute declaration/definition
`NODE_BODY_ATTRIBUTES` -> attribute usage

Comment (sunag) [reply examples/jsm/renderers/nodes/core/NodeBuilder.js:305]: Yes, that's right.

A vertex shader example:
```
#version 450

layout(location = 0) in vec3 position;

NODE_HEADER_ATTRIBUTES
NODE_HEADER_UNIFORMS

void main(){
	NODE_BODY_ATTRIBUTES
	gl_Position = cameraUniforms.projectionMatrix * modelUniforms.modelViewMatrix * vec4( position, 1.0 );
}
```

**Comment by Mugen87** [comment]
@mrdoob Are you okay with merging this? 😊

**Comment by mrdoob** [comment]
Thanks!
